src/main/java/org/alfresco/repo/site/SiteGroupCannedQuery.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * #%L
+ * Alfresco Repository
+ * %%
+ * Copyright (C) 2005 - 2016 Alfresco Software Limited
+ * %%
+ * This file is part of the Alfresco software.
+ * If the software was purchased under a paid Alfresco license, the terms of
+ * the paid license agreement will prevail.  Otherwise, the software is
+ * provided under the following open source license terms:
+ *
+ * Alfresco is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Alfresco is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Alfresco. If not, see <http://www.gnu.org/licenses/>.
+ * #L%
+ */
+package org.alfresco.repo.site;
+
+import org.alfresco.query.AbstractCannedQuery;
+import org.alfresco.query.CannedQueryParameters;
+import org.alfresco.query.CannedQuerySortDetails;
+import org.alfresco.query.CannedQuerySortDetails.SortOrder;
+import org.alfresco.service.cmr.security.AuthorityService;
+import org.alfresco.service.cmr.site.SiteInfo;
+import org.alfresco.service.cmr.site.SiteService;
+import org.alfresco.service.cmr.site.SiteService.SiteMembersCallback;
+import org.alfresco.util.Pair;
+
+import java.util.*;
+
+// TODO currently have to read all sites into memory for sorting purposes. Find a way that doesn't
+public class SiteGroupCannedQuery extends AbstractCannedQuery<SiteMembership>
+{
+  private AuthorityService authorityService;
+  private SiteService siteService;
+
+  protected SiteGroupCannedQuery(
+      SiteService siteService,
+      AuthorityService authorityService,
+      CannedQueryParameters parameters) {
+    super(parameters);
+    this.authorityService = authorityService;
+    this.siteService = siteService;
+  }
+
+  @Override
+  protected List<SiteMembership> queryAndFilter(CannedQueryParameters parameters) {
```

Comment:
As @gavincornwell mentioned below, we should distinguish between SiteUserMembership and SiteGroupMembership as both have different fields. So this SiteGroupCannedQuery should return a List of SiteGroupMembership